File: node-bindings/classic-node/__test__/version_registry_conformance.ts
import { mkdtemp, readFile, readdir, rm, writeFile } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join } from "node:path";
import * as classic from "../index.js";

type JsonObject = Record<string, any>;

/** Observe public metadata and matching against fixture-owned YAML in a dedicated process. */
export async function observeVersionRegistry(fixture: JsonObject): Promise<JsonObject> {
  if (Object.keys(fixture).sort().join() !== "operation,registryYaml,request"
      || !["lookup", "match", "enumerate", "crashgen", "xse", "details"].includes(fixture.operation) || typeof fixture.registryYaml !== "string") {
    throw new Error("unsupported version registry fixture");
  }
  const root = await mkdtemp(join(tmpdir(), "classic-version-registry-conformance-"));
  const previous = process.cwd();
  try {
    await writeFile(join(root, "CLASSIC Main.yaml"), fixture.registryYaml, "utf8");
    // A family has one dedicated serial runner; identical fixture bytes seed OnceLock
    // on first public access without consulting the user's installed metadata.
    process.chdir(root);
    const observation = observe(fixture);
    const files: JsonObject[] = [];
    for (const entry of await readdir(root, { withFileTypes: true })) {
      if (!entry.isFile()) throw new Error("unexpected non-file in version registry workspace");
      files.push({ path: entry.name, content: await readFile(join(root, entry.name), "utf8") });
    }
    return { ...observation, files: files.sort((a, b) => a.path.localeCompare(b.path)) };
  } finally {
    process.chdir(previous);
    await rm(root, { recursive: true, force: true });
  }
}

/** Convert native success/error carriers without inferring domain results from inputs. */
function observe(fixture: JsonObject): JsonObject {
    const request = fixture.request;
    if (fixture.operation === "details") {
      let byVersion;
      try {
        byVersion = classic.getVersionByVersionString(request.version);
      } catch (failure) {
        if (!(failure instanceof Error) || !failure.message.startsWith("Invalid version string:")) throw failure;
        return { result: null, error: { code: "invalid_version" } };
      }
      const handling = classic.getUnknownVersionHandling();
      return { result: {
        byVersion: byVersion?.id ?? null,
        byShortName: classic.getVersionByShortName(request.shortName)?.id ?? null,
        correctIds: classic.getCorrectVersions(request.isVr).map(info => info.id).sort(),
        wrongIds: classic.getWrongVersions(request.isVr).map(info => info.id).sort(),
        addressLibrary: classic.getAddressLibraryFilename(request.version, request.isVr) ?? null,
        crashgenVersions: classic.getCrashgenVersionStrings(request.id),
        exeHashes: classic.getAllExeHashes(request.game, request.isVr).sort(),
        scriptHashes: classic.getAllScriptHashes(request.game, request.isVr),
        versionScriptHashes: classic.getScriptHashesForVersion(request.id),
        strategy: handling.strategy, logLevel: handling.logLevel,
        defaultId: classic.getUnknownVersionDefault(request.game) ?? null,
        compatible: classic.isVersionCompatible(request.id, request.version),
        snapshotIds: classic.getVersionRegistry(request.game, request.isVr).versions.map(info => info.id),
      }, error: null };
    }
    if (fixture.operation === "enumerate") {
      const ids = classic.getAllVersions().map(info => info.id).sort();
      const filteredIds = classic.getAllVersionsForGame(request.game, request.isVr).map(info => info.id).sort();
      return { result: { ids, count: ids.length, filteredIds }, error: null };
    }
    if (fixture.operation === "crashgen") {
      const configs = classic.getCrashgenVersions(request.id).map(crashgen);
      const selected = classic.getCrashgenForVersion(request.id, request.version);
      return { result: { configs, selected: selected == null ? null : crashgen(selected) }, error: null };
    }
    if (fixture.operation === "xse") {
      const xse = classic.getVersionById(request.id)?.xse;
      return { result: xse == null ? null : { acronym: xse.acronym, fullName: xse.fullName,
        compatibleVersion: xse.compatibleVersion, loader: xse.loader, fileCount: xse.fileCount }, error: null };
    }
    if (fixture.operation === "lookup") {
      const info = classic.getVersionById(request.id);
      const result = info == null ? null : {
        id: info.id, version: info.version, shortName: info.shortName,
        game: info.game, docsName: info.docsName, steamId: info.steamId, isVr: info.isVr,
      };
      return { result, error: null };
    }
    try {
      const matched = classic.matchVersion(request.version, request.game, request.isVr);
      return { result: { matchedId: matched.versionInfo?.id ?? null, confidence: matched.confidence, message: matched.message }, error: null };
    } catch (failure: any) {
      // Only the native version parser's documented error earns domain-error credit.
      if (!(failure instanceof Error) || !failure.message.startsWith("Invalid version string:")) throw failure;
      return { result: null, error: { code: "invalid_version" } };
    }
}

/** Project the common native configuration fields without recomputing registry decisions. */
function crashgen(config: JsonObject): JsonObject {
  return { version: config.version, name: config.name, acronym: config.acronym,
    dllFile: config.dllFile, description: config.description, downloadUrl: config.downloadUrl };
}

File: python-bindings/tests/version_registry_conformance.py
"""Observe fixture-seeded Version Registry metadata through actual Python bindings."""

from __future__ import annotations

import os
import tempfile
from collections.abc import Mapping
from pathlib import Path
from typing import Any


def observe_version_registry(fixture: Mapping[str, Any]) -> dict[str, Any]:
    """Invoke lookup/matching in a dedicated serial process and restore its cwd.

    Every fixture carries identical YAML because the public registry initializes
    once per process. Unrelated adapter exceptions propagate as execution failures.
    """
    import classic_version_registry

    if set(fixture) != {"operation", "registryYaml", "request"} or fixture[
        "operation"
    ] not in {"lookup", "match", "enumerate", "crashgen", "xse", "details", "values"}:
        raise ValueError("unsupported version registry fixture")
    previous = Path.cwd()
    with tempfile.TemporaryDirectory(
            prefix="classic-version-registry-conformance-"
    ) as directory:
        root = Path(directory)
        (root / "CLASSIC Main.yaml").write_bytes(
            fixture["registryYaml"].encode("utf-8")
        )
        try:
            # Isolate first-use YAML discovery from repository/user configuration.
            os.chdir(root)
            observation = _observe(fixture, classic_version_registry.VersionRegistry())
            if (
                    _observe(fixture, classic_version_registry.get_version_registry())
                    != observation
            ):
                raise ValueError(
                    "registry constructor and singleton entry point disagree"
                )
            files = []
            for path in sorted(root.iterdir()):
                if path.is_symlink() or not path.is_file():
                    raise ValueError(
                        "unexpected non-file in version registry workspace"
                    )
                files.append(
                    {"path": path.name, "content": path.read_bytes().decode("utf-8")}
                )
            return {**observation, "files": files}
        finally:
            os.chdir(previous)


def _observe(fixture: Mapping[str, Any], registry: Any) -> dict[str, Any]:
    """Convert native success/error carriers without inferring results from inputs."""
    request = fixture["request"]
    if fixture["operation"] == "values":
        info = registry.get_by_id(request["id"])
        clone = registry.get_by_id(request["id"])
        other = registry.get_by_id(request["otherId"])
        bounds = info.compatible_range
        crashgen = info.get_crashgen_for_version(request["crashgen"])
        return {
            "result": {
                "range": [bounds.min_version, bounds.max_version],
                "contains": [
                    bounds.contains(version) for version in request["versions"]
                ],
                "versionCompatible": [
                    info.is_compatible_with(version) for version in request["versions"]
                ],
                "crashgenCompatible": [
                    crashgen.is_compatible_with(version)
                    for version in request["versions"]
                ],
                "compatibleCrashgens": [
                    [
                        config.version
                        for config in info.get_compatible_crashgens(version)
                    ]
                    for version in request["versions"]
                ],
                "defaultCrashgens": [
                    config.version for config in info.get_compatible_crashgens()
                ],
                "versions": info.get_crashgen_version_strings(),
                "selected": crashgen.version,
                "missing": info.get_crashgen_for_version("missing"),
                "equalClone": info == clone,
                "differentId": info == other,
                "hashClone": hash(info) == hash(clone),
            },
            "error": None,
        }
    if fixture["operation"] == "details":
        try:
            by_version = registry.get_by_version(request["version"])
        except ValueError as failure:
            if not str(failure).startswith("Invalid version: Invalid version string:"):
                raise
            return {"result": None, "error": {"code": "invalid_version"}}
        by_name = registry.get_by_short_name(request["shortName"])
        info = registry.get_by_id(request["id"])
        handling = registry.unknown_version_handling
        if (
                info is not None
                and info.address_library is not None
                and info.address_library.filename
                != registry.get_address_library_filename(info.version, info.is_vr)
        ):
            raise ValueError(
                "address-library carrier and public filename query disagree"
            )
        if (
                info is not None
                and info.get_crashgen_version_strings()
                != registry.get_crashgen_versions(request["id"])
        ):
            raise ValueError("registry and value crashgen version strings disagree")
        return {
            "result": {
                "byVersion": None if by_version is None else by_version.id,
                "byShortName": None if by_name is None else by_name.id,
                "correctIds": sorted(
                    info.id for info in registry.get_correct_versions(request["isVr"])
                ),
                "wrongIds": sorted(
                    info.id for info in registry.get_wrong_versions(request["isVr"])
                ),
                "addressLibrary": registry.get_address_library_filename(
                    request["version"], request["isVr"]
                ),
                "crashgenVersions": registry.get_crashgen_versions(request["id"]),
                "exeHashes": sorted(
                    registry.get_all_exe_hashes(request["game"], request["isVr"])
                ),
                "scriptHashes": {
                    key: sorted(values)
                    for key, values in registry.get_all_script_hashes(
                        request["game"], request["isVr"]
                    ).items()
                },
                "versionScriptHashes": registry.get_script_hashes_for_version(
                    request["id"]
                ),
                "strategy": handling.strategy,
                "logLevel": handling.log_level,
                "defaultId": handling.get_default(request["game"]),
                "compatible": info is not None
                              and info.is_compatible_with(request["version"]),
                "snapshotIds": [
                    info.id
                    for info in registry.get_all_for_game(
                        request["game"], request["isVr"]
                    )
                ],
            },
            "error": None,
        }
    if fixture["operation"] == "enumerate":
        ids = sorted(info.id for info in registry.get_all())
        filtered = sorted(
            info.id
            for info in registry.get_all_for_game(request["game"], request["isVr"])
        )
        return {
            "result": {"ids": ids, "count": len(ids), "filteredIds": filtered},
            "error": None,
        }
    if fixture["operation"] == "crashgen":
        configs = [
            _crashgen(config) for config in registry.get_crashgen_configs(request["id"])
        ]
        selected = registry.get_crashgen_for_version(request["id"], request["version"])
        info = registry.get_by_id(request["id"])
        if info is not None:
            value_selected = info.get_crashgen_for_version(request["version"])
            if (None if selected is None else _crashgen(selected)) != (
                    None if value_selected is None else _crashgen(value_selected)
            ):
                raise ValueError("registry and value crashgen selection disagree")
        return {
            "result": {
                "configs": configs,
                "selected": None if selected is None else _crashgen(selected),
            },
            "error": None,
        }
    if fixture["operation"] == "xse":
        info = registry.get_by_id(request["id"])
        xse = None if info is None else info.xse
        result = (
            None
            if xse is None
            else {
                "acronym": xse.acronym,
                "fullName": xse.full_name,
                "compatibleVersion": xse.compatible_version,
                "loader": xse.loader,
                "fileCount": xse.file_count,
            }
        )
        return {"result": result, "error": None}
    if fixture["operation"] == "lookup":
        info = registry.get_by_id(request["id"])
        result = (
            None
            if info is None
            else {
                "id": info.id,
                "version": info.version,
                "shortName": info.short_name,
                "game": info.game,
                "docsName": info.docs_name,
                "steamId": info.steam_id,
                "isVr": info.is_vr,
            }
        )
        return {"result": result, "error": None}
    try:
        matched = registry.match_version(
            request["version"], request["game"], request["isVr"]
        )
        import classic_version_registry

        convenience = classic_version_registry.match_version_string(
            request["version"], request["game"], request["isVr"]
        )
        if _match_result(matched) != _match_result(convenience):
            raise ValueError("registry matching entry points disagree")
    except ValueError as failure:
        if not str(failure).startswith("Invalid version: Invalid version string:"):
            raise
        return {"result": None, "error": {"code": "invalid_version"}}
    return {"result": _match_result(matched), "error": None}


def _match_result(matched: Any) -> dict[str, Any]:
    """Compare both Python matching entry points through native result fields."""
    confidence = matched.confidence_enum
    copy = matched.confidence_enum
    if confidence != copy or hash(confidence) != hash(copy):
        raise ValueError("confidence copies disagree in equality or hash")
    if confidence.is_high_confidence() != (str(confidence) in {"exact", "range"}):
        raise ValueError("confidence predicate disagrees with public confidence name")
    if str(confidence) != matched.confidence:
        raise ValueError("confidence object and public string disagree")
    return {
        "matchedId": None if matched.version_info is None else matched.version_info.id,
        "confidence": str(matched.confidence),
        "message": matched.message,
    }


def _crashgen(config: Any) -> dict[str, Any]:
    """Project common crash generator metadata from the native object."""
    return {
        "version": config.version,
        "name": config.name,
        "acronym": config.acronym,
        "dllFile": config.dll_file,
        "description": config.description,
        "downloadUrl": config.download_url,
    }
